miniapps/meshing/shaper.cpp

```diff
@@ -0,0 +1,177 @@
+   // Rescaling to [-1,1]^sdim
+   for (int i = 0; i < x.Size(); i++)
+   {
+      x(i) = (2*x(i)-xmin(i)-xmax(i))/(xmax(i)-xmin(i));
+   }
+
+   // A simple annulus/shell
+   if (x.Normlp(p) > 0.4 && x.Normlp(p) < 0.6) { return 1; }
+   if (x.Normlp(p) < 0.4 || x.Normlp(p) > 0.6) { return 2; }
+   return 0;
+}
+
+int main (int argc, char *argv[])
+{
+   int sd = 2;
+   int nclimit = 1;
+   const char *mesh_file = "../../data/inline-quad.mesh";
+
+   // Parse command line
+   OptionsParser args(argc, argv);
+   args.AddOption(&mesh_file, "-m", "--mesh",
+                  "Input mesh file to shape materials in.");
+   args.AddOption(&sd, "-sd", "--sub-divisions",
+                  "Number of element subdivisions for interface detection.");
+   args.AddOption(&nclimit, "-ncl", "--nc-limit",
+                  "Level of hanging nodes allowed (-1 = unlimited).");
+   args.Parse();
+   if (!args.Good()) { args.PrintUsage(cout); return 1; }
+   args.PrintOptions(cout);
+
+   // Read initial mesh, get dimensions and bounding box
+   Mesh mesh(mesh_file, 1, 1);
+   int dim = mesh.Dimension();
+   int sdim = mesh.SpaceDimension();
+   Vector xmin, xmax;
+   mesh.GetBoundingBox(xmin, xmax);
+
+   // NURBS meshes don't support non-conforming refinement for now
+   if (mesh.NURBSext) { mesh.SetCurvature(2); }
+
+   // Mesh attributes will be visualized as piece-wise constants
+   L2_FECollection attr_fec(0, dim);
+   FiniteElementSpace attr_fespace(&mesh, &attr_fec);
+   GridFunction attr(&attr_fespace);
+
+   // GLVis server to visualize to
+   char vishost[] = "localhost";
+   int  visport   = 19916;
+   socketstream sol_sock(vishost, visport);
+   sol_sock.precision(8);
+
+   // Shaping loop
+   for (int iter = 0; 1; iter++)
+   {
+      Array<Refinement> refs;
+      for (int i = 0; i < mesh.GetNE(); i++)
+      {
+         bool refine = false;
+
+         // Sample materials in each element using "sd" sub-divisions
+         Vector pt;
+         int geom = mesh.GetElementBaseGeometry(i);
+         ElementTransformation *T = mesh.GetElementTransformation(i);
+         RefinedGeometry *RefG = GlobGeometryRefiner.Refine(geom, sd, 1);
+         IntegrationRule &ir = RefG->RefPts;
+
+         // Refine any element where different materials are detected. A more
+         // sophisticated logic can be implemented here -- e.g. don't refine
+         // the interfaces between certain materials.
+         int mat = 0;
+         double matsum = 0.0;
+         for (int j = 0; j < ir.GetNPoints(); j++)
+         {
+            T->Transform(ir.IntPoint(j), pt);
+            mat = material(pt, xmin, xmax);
+            matsum += mat;
+            if ((int)matsum != mat*(j+1))
+            {
+               refine = true;
+            }
+         }
+
+         // Set the element attribute as the "average". Other choices are
+         // possible here too, e.g. attr(i) = mat;
+         attr(i) = (int)(matsum/ir.GetNPoints());
+         mesh.SetAttribute(i, attr(i));
```

Comment:
Is this for visualization only? This choice does not affect refinement, does it?